Documentation Update (#271)
OpenLiberty/liberty-tools-vscode: docs/user-guide.md

## Before you begin

### Software requirements
<!-- Match the VSCode and Java version with the README.md -->
- **Visual Studio Code 1.78.0** or later.
- **Java 17** or later is required by Liberty Tools for Visual Studio Code. To point Visual Studio Code to a specific runtime, see [Settings](#settings) for more information.

### Application requirements

- Define a Liberty `server.xml` configuration file in the `src/main/liberty/config` location.
- [Configure the Liberty Maven Plugin](https://github.com/OpenLiberty/ci.maven#configuration) or [configure the Liberty Gradle Plugin](https://github.com/OpenLiberty/ci.gradle#adding-the-plugin-to-the-build-script). We recommend using the most recent versions of the plugins as they include several important bug fixes.

    - [Liberty Maven Plugin latest release](https://github.com/OpenLiberty/ci.maven/releases/latest)

    - [Liberty Gradle Plugin latest release](https://github.com/OpenLiberty/ci.gradle/releases/latest)

### Configure your Java

> Liberty Tools for Visual Studio Code requires Java 17 or later to ensure that the [Liberty Config Language Server](https://github.com/OpenLiberty/liberty-language-server), the [Eclipse Language Server for Jakarta EE](https://github.com/eclipse/lsp4jakarta), the [XML Language Server](https://github.com/eclipse/lemminx) and start properly. A toast message alerts you if any language server fails to run or if Visual Studio Code has trouble locating your Java.

Reviewer: and the [XML Language Server](https://github.com/eclipse/lemminx) start properly.

Reviewer 2: you might consider putting the language servers into a list for readability
```suggestion
Liberty Tools for Visual Studio Code requires Java 17 or later to ensure that the following underlying language servers start properly:
- [Liberty Config Language Server](https://github.com/OpenLiberty/liberty-language-server)
- [Eclipse Language Server for Jakarta EE](https://github.com/eclipse/lsp4jakarta)
- [XML Language Server](https://github.com/eclipse/lemminx) 

A toast message alerts you if any language server fails to run or if Visual Studio Code has trouble locating your Java.
```

not sure if "underlying" is an accurate characterization, edit as needed

<img src="/docs/screenshots/update%20jdk%20toast.png" width="40%"/> 
<img src="/docs/screenshots/java_17_toast_alert.png" width="40%" />

To resolve this issue, you can define properties in your Visual Studio Code [settings.json](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson) file or set system environment variables to point Liberty Tools to your Java 17 or later. 

For both Liberty Config Language Server and Eclipse Language Server for Jakarta EE, Liberty Tools for Visual Studio Code will check for the Java versions in the following order:
1. The [embedded JRE](https://github.com/redhat-developer/vscode-java#java-tooling-jdk) included by [Language Support for Java(TM) by Red Hat](https://marketplace.visualstudio.com/items?itemName=redhat.java). 

>   - By default, Liberty Tools installs the latest version of the rehat.java plugin, which contains an embedded JRE higher than Java 17 and therefore no configuration is required. However, if using an older version of the redhat.java plugin or using the universal version without the embedded JRE, use `java.jdt.ls.java.home` to specify your new JDK path.

Reviewer: "rehat.java", "redhat.java" should these be expanded to the full name of the plugin?

2. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` will take precedence)

> For the XML Language Server, or LemMinX, Liberty Tools for Visual Studio Code will check for the Java versions in the following order:

Reviewer: Will people know about the LemMinX component? Could we say "the language server supplied by the XML Language Support extension." Same comment for the mention of LemMinX in the table.

Author: <img width="802" alt="image" src="https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/33f2b4a4-07a0-447b-8b8b-9e718f807ec8">

I've appended LemMinX onto the reference to the server here. Going forward, I think we can start referencing to it as LemMinX.

Reviewer: Ok, that clarifies. Thanks.

1. `xml.java.home` in settings.json

> 2. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` will take precedence)

Reviewer 2: ```suggestion
2. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` takes precedence)
```

![settings.json example](/docs/screenshots/settings.json%20path%20example.png)

### External settings

The following settings, which are provided by external extensions, are honoured when you run dev mode commands.

| Setting | Description | Provided By |
| --- | --- | --- |

> | `maven.executable.path` | Maven commands executed by dev mode will honour this setting. When this value is empty, it tries to use `mvn` or `mvnw` according to the value of `maven.executable.preferMavenWrapper`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |

Reviewer 2: ```suggestion
| `maven.executable.path` | Maven commands executed by dev mode honour this setting. When this value is empty, dev mode tries to use `mvn` or `mvnw` according to the value of `maven.executable.preferMavenWrapper`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |
```

> | `maven.executable.preferMavenWrapper` | Maven commands executed by dev mode will honour this setting. If true, it tries to use `mvnw` if a Maven wrapper file can be found. Otherwise it will use `mvn`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |

Reviewer 2: ```suggestion
| `maven.executable.preferMavenWrapper` | Maven commands executed by dev mode honour this setting. If this setting set to `true`, dev mode tries to use `mvnw` if a Maven wrapper file can be found. Otherwise, it uses `mvn`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |
```

> | `java.import.gradle.wrapper.enabled` | Gradle commands executed by dev mode will honour this setting. If true, it tries to use `gradlew` if a Gradle wrapper file can be found. Otherwise it will use `gradle`. | [Language support for Java extension](https://marketplace.visualstudio.com/items?itemName=redhat.java) |

Reviewer 2: ```suggestion
| `java.import.gradle.wrapper.enabled` | Gradle commands executed by dev mode honour this setting. If this setting is set to `true`, dev mode tries to use `gradlew` if a Gradle wrapper file can be found. Otherwise, it uses `gradle`. | [Language support for Java extension](https://marketplace.visualstudio.com/items?itemName=redhat.java) |
```

> | `xml.java.home` | This property allows a user to define their LemMinX language server runtime without altering the `JAVA_HOME` environment variable.  | Not set |

Reviewer 2: ```suggestion
| `xml.java.home` | This setting allows a user to define their LemMinX language server runtime without altering the `JAVA_HOME` environment variable.  | Not set |
```

for consistency- alternatively, if all settings are properties, can replace mentions in this table of "setting" with "property"

> | `java.jdt.ls.java.home` | Specifies the folder path to the JDK (17 or more recent) used to launch the Java Language Server. This setting will replace the Java extension's embedded JRE to start the Java Language Server.  | Not set |

Reviewer 2: ```suggestion
| `java.jdt.ls.java.home` | This setting specifies the folder path to the JDK (17 or later) that is used to launch the Java Language Server. This setting replaces the Java extension's embedded JRE to start the Java Language Server.  | Not set |
```

## Before you begin

### Software requirements
<!-- Match the VSCode and Java version with the README.md -->
- **Visual Studio Code 1.78.0** or later.
- **Java 17** or later is required by Liberty Tools for Visual Studio Code. To point Visual Studio Code to a specific runtime, see [Settings](#settings) for more information.

### Application requirements

- Define a Liberty `server.xml` configuration file in the `src/main/liberty/config` location.
- [Configure the Liberty Maven Plugin](https://github.com/OpenLiberty/ci.maven#configuration) or [configure the Liberty Gradle Plugin](https://github.com/OpenLiberty/ci.gradle#adding-the-plugin-to-the-build-script). We recommend using the most recent versions of the plugins as they include several important bug fixes.

    - [Liberty Maven Plugin latest release](https://github.com/OpenLiberty/ci.maven/releases/latest)

    - [Liberty Gradle Plugin latest release](https://github.com/OpenLiberty/ci.gradle/releases/latest)

> ### Configure your Java

Reviewer 3: ```suggestion
### Configure Java runtime for language servers
```

Liberty Tools for Visual Studio Code requires Java 17 or later to ensure that the following language servers start properly:
- [Liberty Config Language Server](https://github.com/OpenLiberty/liberty-language-server)
- [Eclipse Language Server for Jakarta EE](https://github.com/eclipse/lsp4jakarta)
- [XML Language Server (LemMinX)](https://github.com/eclipse/lemminx) 

A toast message alerts you if any language server fails to run or if Visual Studio Code has trouble locating your Java.

<img src="/docs/screenshots/update%20jdk%20toast.png" width="40%"/> 
<img src="/docs/screenshots/java_17_toast_alert.png" width="40%" />

> To resolve this issue, you can define properties in your Visual Studio Code [settings.json](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson) file or set system environment variables to point Liberty Tools to your Java 17 or later. 

Reviewer 3: ```suggestion
To resolve this issue, you can define settings in your Visual Studio Code [settings.json](https://code.visualstudio.com/docs/getstarted/settings#_settingsjson) file or set system environment variables to point Liberty Tools to Java 17 or later. 
```

For both Liberty Config Language Server and Eclipse Language Server for Jakarta EE, Liberty Tools for Visual Studio Code will check for the Java versions in the following order:

> 1. The [embedded JRE](https://github.com/redhat-developer/vscode-java#java-tooling-jdk) included by [Language Support for Java(TM) by Red Hat](https://marketplace.visualstudio.com/items?itemName=redhat.java). 

Reviewer 3: Based on the latest info, is the embedded JRE still number 1?
Wondering how we should structure this section now...

>   - By default, Liberty Tools installs the latest version of the Language Support for Java(TM) by Red Hat plugin. The latest version contains an embedded JRE higher than Java 17 and therefore no additional configuration is required. However, using an older version of the Language Support for Java(TM) by Red Hat plugin or using the universal version without the embedded JRE causes an issue, use `java.jdt.ls.java.home` to specify your new JDK path.

Reviewer 3: ```suggestion
  - By default, Liberty Tools installs the latest version of the Language Support for Java(TM) by Red Hat extension. The latest version contains an embedded JRE higher than Java 17 and therefore no additional configuration is required. However, if using an older version of the Language Support for Java(TM) by Red Hat plugin or using the universal version without the embedded JRE causes an issue, one of the options below must be configured to use Java 17 or later.
  2. `java.jdt.ls.java.home` in settings.json
```

> 2. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` will take precedence)

Reviewer 3: ```suggestion
3. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` will take precedence)
```

For LemMinX, Liberty Tools for Visual Studio Code will check for the Java versions in the following order:
1. `xml.java.home` in settings.json
2. `JDK_HOME` or `JAVA_HOME` as system environment variables. (Note: if both `JDK_HOME` and `JAVA_HOME` are set, `JDK_HOME` takes precedence)

![settings.json example](/docs/screenshots/settings.json%20path%20example.png)

> ### External settings

Reviewer 3: should we use the title "External extension settings"?

> The following settings, which are provided by external extensions, are honoured when you run dev mode commands.

Reviewer 3: I think we should remove this line. `xml.java.home` and `java.jdt.ls.java.home` are not connected to dev mode.

| --- | --- | --- |
| `maven.executable.path` | Maven commands executed by dev mode honour this setting. When this value is empty, dev mode tries to use `mvn` or `mvnw` according to the value of `maven.executable.preferMavenWrapper`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |
| `maven.executable.preferMavenWrapper` | Maven commands executed by dev mode honour this setting. If this setting set to `true`, dev mode tries to use `mvnw` if a Maven wrapper file can be found. Otherwise, it uses `mvn`. | [Maven for Java extension](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-maven) |
| `java.import.gradle.wrapper.enabled` | Gradle commands executed by dev mode honour this setting. If this setting is set to `true`, dev mode tries to use `gradlew` if a Gradle wrapper file can be found. Otherwise, it uses `gradle`. | [Language support for Java extension](https://marketplace.visualstudio.com/items?itemName=redhat.java) |
| `xml.java.home` | This setting allows a user to define their LemMinX language server runtime without altering the `JAVA_HOME` environment variable.  | Not set |

> | `java.jdt.ls.java.home` | This setting specifies the folder path to the JDK (17 or later) that is used to launch the Java Language Server. This setting replaces the Java extension's embedded JRE to start the Java Language Server.  | Not set |

Reviewer 3: ```suggestion
| `java.jdt.ls.java.home` | This setting specifies the folder path to the JDK (17 or later) that is used to launch the Java Language Server. This setting replaces the Java extension's embedded JRE to start the Java Language Server as well as the Liberty Config Language Server and the Eclipse Language Server for Jakarta EE  | Not set |
```
